### DISCUSSION TITLE
Ancestry

### DISCUSSION THREAD

**Discussion by rctneil** [avo-hq/avo#520] [Help]
Hi,

I just installed this on my new Family Tree app that uses [Ancestry](https://github.com/stefankroes/ancestry) but am finding it difficult to integrate it with the hierarchy of records.

I need to be able to choose a Person's parent and also want to show a Person's list of Children too.

I may be able to use the select field options to output a list of people to choose the parent person from, but when saving the record, it needs to use Ancestry's features in order to save the hierarchy correctly.

Any suggestions as to how I can do this?

**Comment by adrianthedev** [comment]
Hmm... What Ancestry does is quite niched. Avo probably doesn't support that kind of feature.

**Comment by rctneil** [comment]
Hi @adrianthedev 

Thanks for your reply.

I understand that a hierarchical view may not be possible, but would it be possible, in some way, to be able to control this using Avo.

At a minimum, the features I would need are:

1. Be able to select a Persons parent and update this correctly. See note below.
2. Be able to view a Persons children on their Show view.  These can be got from @person.children.   Would there be a way to use the existing associations views to display this?

Note: I tried using the Select field type to allow a new Parent to be chosen but the update doesn't work.  My code for t is as follows:

```
    field :parent, as: :select, options: ->(model: m, resource: r, view: v, field: f) {
        Person.all.map { |p| [p.fullname, p.id] }
    }, placeholder: 'Choose parent'
```

I really would like to use Avo with this project but I really could do with being able to add the two features above.  Maybe they are already possible but I just need a little help connecting the dots etc to enable it.

**Reply by adrianthedev** [reply]
I see what you mean. 
For 1. the select might not work on its own because it's trying to set the actual attribute of that model.
I'm spitballing here, but you might be able to catch the value of `field :parent` from your example in a model callback. Maybe `before_save`? And set the actual parent there using ancestry?

For 2. You should be able to create a new [custom field](https://docs.avohq.io/1.0/custom-fields.html) (make sure you [change the license type to `pro`](https://docs.avohq.io/1.0/licensing.html#add-the-license-key)).
You'll get a view component that you can customize to your liking and generate the tree.

Please let me know if this helped.

**Comment by rctneil** [comment]
Hi, @adrianthedev 

Is there a way for me to trial this in advance of purchasing a Pro license?  I need to ensure it will work first before I commit to purchasing anything.

**Reply by adrianthedev** [reply]
Yes. On local, you can change the `config.license = 'pro'` in the initializer. All the pro features will work while in the `development` env.

For staging/production, I'll send you a trial key right now.

**Comment by rctneil** [comment]
Hi Adrian,

Many thanks for that.  I'll give it a go.  I'm not super fast with Rails or tremendously great at it but i'll give it a bash.

I just updated the field name for :parent to :parent_id.

This now makes the Select I posted above show the correct parent when on the Edit view. Progress!    I then tried making a change to the parent and clicking "Save".   I get the below error, but then if i refresh the page, it did actually manage to update the parent successfully.

Any idea what may be contributing to the error?

Showing /Users/<user>/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/avo-1.13.3/app/views/avo/partials/_turbo_frame_wrap.html.erb where line #2 raised:

```
undefined method `invert' for #<Array:0x00007f9b86045ed8>
Did you mean?  insert

return value if display_value

          computed_options.invert.stringify_keys[value]
        elsif enum.present?
          if display_value
            options[value]
```

I hope to try the custom field thing later this week or next.     As I don;t need to build a tree structure for children, but just a simple list of associated immediately related children, would I be able to use the listing styles that you have for the association functionality.  I basically want the children of a person to be displayed in a form similar to the below from your documentation:

![image](https://user-images.githubusercontent.com/774459/137905609-ed9f6962-0e45-4245-9f29-0113ffca98b0.png)

Really appreciate your help so far!

Thanks,
Neil

**Reply by adrianthedev** [reply]
From the error message it seems it's trying to apply the invert method to an `Array`, and that method does not exist on an `Array`. The `select` field does not take an array as the value but a `Hash`.

So from what you posted earlier (you were very close), you should apply this change to turn that array to a Hash.

```diff
field :parent, as: :select, options: ->(model: m, resource: r, view: v, field: f) {
-  Person.all.map { |p| [p.fullname, p.id] }
+  Person.all.map { |p| [p.fullname, p.id] }.to_h
}, placeholder: 'Choose parent'
```

Let me know if that helped.

**Reply by rctneil** [reply]
@adrianthedev That's perfect!  That works perfectly now.   Next step is to try creating a custom field.  I'll report back!

**Comment by rctneil** [comment]
@adrianthedev Hi,

I've started to create the custom field but have two queries:

Currently, my (admittedly very basic) custom filed looks like:

```
<%= show_field_wrapper field: @field, index: @index do %>
  <% @field.value.each do |child| %>
    <div>
      <%= link_to child.fullname, resources_person_path(child) %>
    </div>
  <% end %>
<% end %>
```

The link_to route does not appear to be working as it cannot find the route? How can I get the route correct so each child name links to their show view in Avo?

Also,  is there an example of markup that I can duplicate in order to ensure the table of children matches the Avo style and layout?
